### PR TITLE
soc: Add support for LPC11U6X soc series

### PR DESCRIPTION
**Pull request by mbittan**
This pull request adds support for NXP LPC11U6X soc series which is embedded in a lot a LaCie/Seagate devices. It includes drivers for clock control, pinmux, gpio, uart, i2c and EEPROM devices and adds support for the NXP LPCXPRESSO11U68 evaluation board based on the LPC11U68JBD100 MCU.

Pinmuxing for I2C and UART devices is handled directly by the corresponding driver instance, which get the pinmux information from the device tree.

This series of patch has been tested on both the LPCXPRESSO11U68 board and another board based on the LPC11U67 MCU.

### CODE REVIEW

**Review by galak — COMMENTED** [review]
Are the GPIO, EEPROM, or I2C bits required to boot the board/SoC?  If not I'd ask we split the PR up and have those changes in separate PRs to make it easier to review.

File: soc/arm/nxp_lpc/lpc11u6x/soc.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @return 0
+ */
+
+static int nxp_lpc11u6x_init(struct device *arg)
+{
+	ARG_UNUSED(arg);
+
+	/* old interrupt lock level */
+	unsigned int oldLevel;
```
Comment (carlescufi) [inline soc/arm/nxp_lpc/lpc11u6x/soc.c:38]: Please no camel case

File: boards/arm/lpcxpresso11u68/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+NXP LPCXpresso11U68
+###################
+
+Overview
+********
+
+The LPCXpresso11u68 development board uses an NXP LPC11U68 MCU provides an
+affordable way to prototype new systems.
+
```
Comment (carlescufi) [inline boards/arm/lpcxpresso11u68/doc/index.rst:11]: Can we remove the marketing speak? Perhaps just mention the MPU

File: boards/arm/lpcxpresso11u68/support/lpc_checksum.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2013 Bas Stottelaar
```
Comment (carlescufi) [inline boards/arm/lpcxpresso11u68/support/lpc_checksum.py:3]: @MaureenHelm is it OK to have this Python script here?

File: drivers/clock_control/clock_control_lpc11u6x.h [hunk 4]
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2020, Seagate Technology LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_CLOCK_CONTROL_CLOCK_CONTROL_LPC11U6X_H_
+#define ZEPHYR_DRIVERS_CLOCK_CONTROL_CLOCK_CONTROL_LPC11U6X_H_
+
+#define LPC11U6X_SYS_AHB_CLK_CTRL_I2C0      (1 << 5)
```
Comment (carlescufi) [inline drivers/clock_control/clock_control_lpc11u6x.h:10]: Can’t all those macros come from mcux?

File: drivers/i2c/i2c_lpc11u6x.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2020, Seagate Technology LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_I2C_I2C_LPC11U6X_H_
+#define ZEPHYR_DRIVERS_I2C_I2C_LPC11U6X_H_
+
+#define LPC11U6X_I2C_CONTROL_AA           (1 << 2)
```
Comment (carlescufi) [inline drivers/i2c/i2c_lpc11u6x.h:10]: Same, can’t we use mcux for this?

**Review by carlescufi — CHANGES_REQUESTED** [review]
Thanks for the contribution! Not an expert on this platform but have a couple of comments and questions.

File: drivers/gpio/Kconfig.lpc11u6x [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# LPC11U6x GPIO configuration options
+
+# Copyright (c) 2020 2020 Seagate Technology LLC
```
Comment (MaureenHelm) [inline drivers/gpio/Kconfig.lpc11u6x:3]: Don't need two 2020's. :)

File: dts/arm/nxp/nxp_lpc11u6x.dtsi [hunk 7]
```diff
@@ -6,6 +6,7 @@
 #include <arm/armv6-m.dtsi>
 #include <dt-bindings/clock/lpc11u6x_clock.h>
 #include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/clock/lpc11u6x_clock.h>
```
Comment (MaureenHelm) [inline dts/arm/nxp/nxp_lpc11u6x.dtsi]: Don't need to add this again

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the contribution!

Please review the failed checks (codeowners, nits)

**Comment by mbittan** [comment]
Thank you everyone for your reviews. I just pushed a new version of the branch that addresses most of the issues that have been raised.
Regarding the license of the script lpc_checksum.py, is there a specific procedure to follow ?

File: drivers/pinmux/Kconfig [hunk 8]
```diff
@@ -30,28 +30,30 @@ config PINMUX_INIT_PRIORITY
...
+source "drivers/pinmux/Kconfig.mcux"
+
+source "drivers/pinmux/Kconfig.mcux_lpc"
+
 source "drivers/pinmux/Kconfig.rv32m1"
 
-source "drivers/pinmux/Kconfig.xec"
+source "drivers/pinmux/Kconfig.sifive"
 
-source "drivers/pinmux/Kconfig.hsdk"
```
Comment (henrikbrixandersen) [inline drivers/pinmux/Kconfig:55]: Could the sorting of this file be submitted as a separate PR?

File: drivers/serial/CMakeLists.txt [hunk 9]
```diff
@@ -10,6 +10,7 @@ zephyr_library_sources_if_kconfig(uart_esp32.c)
 zephyr_library_sources_if_kconfig(uart_sifive.c)
 zephyr_library_sources_if_kconfig(uart_gecko.c)
 zephyr_library_sources_if_kconfig(leuart_gecko.c)
+zephyr_library_sources_ifdef(CONFIG_UART_LPC11U6X uart_lpc11u6x.c)
```
Comment (henrikbrixandersen) [inline drivers/serial/CMakeLists.txt:13]: Why not use `zephyr_library_sources_if_kconfig()`?

File: drivers/serial/Kconfig.lpc11u6x [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+config UART_LPC11U6X_PORT3
+	bool "Enable LPC11U6X UART3"
+	help
+	  Enable UART port 3
+
+config UART_LPC11U6X_PORT4
+	bool "Enable LPC11U6X UART4"
+	help
+	  Enable UART port 4
```
Comment (henrikbrixandersen) [inline drivers/serial/Kconfig.lpc11u6x:39]: Could these be obtained from devicetree instead of Kconfig?

File: drivers/serial/uart_lpc11u6x.c [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+		ret |= UART_BREAK;
+	}
+
+	return ret;
+}
+
+static int lpc11u6x_uart0_configure(struct device *dev,
+				    const struct uart_config *cfg)
+{
+	return -ENOTSUP;
```
Comment (henrikbrixandersen) [inline drivers/serial/uart_lpc11u6x.c:67]: Wont this give problems with portable code, which attempts to configure the UART? Would it be better to check the given configuration and only return `-ENOTSUP` if the requested configuration is not applicable for the LPC11U6x?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Thank you for this large contribution. It looks really good.

I am a bit divided regarding the split between `soc.c` and the clock controller driver, but since the clock controller driver is very SoC specific, this might be the best solution.

The pinmux design and implementation looks very interesting. I would be very interested in hearing @galak if this is in-line with the generic pinmux design plans for Zephyr v2.4 (and if so, if we should start migrating other SoCs/drivers towards this).

File: drivers/eeprom/eeprom_lpc11u6x.c [hunk 12]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_lpc11u6x_eeprom
```
[the remark this answers is not in the document]
Comment (simonguinot) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: Hi Henrik,
I don't know how much the IAP is shared between all the LPCs. Depending on the MCUs some feature may be available or not. The entry codes and the return status may be different as well. I think @MaureenHelm is the one able to answer this.

File: drivers/gpio/Kconfig.lpc11u6x [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+config GPIO_LPC11U6X_PORT1
+	bool "Port 1"
+	help
+	  Enable port 1.
+
+config GPIO_LPC11U6X_PORT2
+	bool "Port 2"
+	help
+	  Enable port 2.
```
[the remark this answers is not in the document]
Comment (simonguinot) [reply drivers/gpio/Kconfig.lpc11u6x:29]: Yes it can. In my understanding (coming from the Linux world), DT is about describing the hardware, and not about configuring it. In Linux there is a strong line about that. So if a board does provide two GPIO ports, then they have to be defined and enabled both in the DT. And the application can choose to use them, or not... With @mbittan we realized this was maybe not the way it was working in Zephyr. But it was pretty late in the process and we chose to keep the GPIO driver with the CONFIG_ port options and an abnormal macros usages to ensure a correct match between GPIO instances and ports numbers. 

So you confirm that the correct way to disable a GPIO port is by disabling its DT node from an application overlay ?

File: drivers/gpio/gpio_lpc11u6x.c [hunk 14]
```diff
@@ -0,0 +1,655 @@
...
+	}
+
+	if (flags & GPIO_OUTPUT_INIT_LOW) {
+		gpio_regs->clr[port_num] |= BIT(pin);
+	}
+
+	/*
+	 * TODO: maybe configure the STARTERP0 register to allow wake-up from
+	 * deep-sleep or power-down modes.
+	 */
```
[the remark this answers is not in the document]
Comment (simonguinot) [reply drivers/gpio/gpio_lpc11u6x.c:180]: I think this should come from a flag when configuring a GPIO interrupt.

File: drivers/gpio/Kconfig.lpc11u6x (hunk 13 above)
[the remark this answers is not in the document]
Comment (pabigot) [reply drivers/gpio/Kconfig.lpc11u6x:29]: Yes, that's how this is being done these days.  It's closer to the Linux way than Kconfig.  Some day we should have the ability to enable/disable at runtime, too, for power saving.

**Comment by henrikbrixandersen** [comment]
> Regarding the license of the script lpc_checksum.py, is there a specific procedure to follow ?

@mbittan Please see https://docs.zephyrproject.org/latest/contribute/index.html#contribute-non-apache

File: drivers/pinmux/Kconfig (hunk 8 above)
Comment (simonguinot) [reply drivers/pinmux/Kconfig:55]: Do you suggest to make this PR depend on another one sorting the Kconfig / CMakeLists.txt files ? Or do you rather suggest to drop this patch and eventually come back "later" with another PR to do some sorting ?

Honestly, this was the occasion to do some cleaning but we can drop this patch...

Comment (pabigot) [reply drivers/pinmux/Kconfig:55]: My personal opinion is that it's OK to mix commits to do cleanup into PRs that subsequently change the file, which is what seems to be happening here.  It's not generally ok to combine cleanup with other changes in the commit , which is not happening here.

But we don't have a policy.

Comment (simonguinot) [reply drivers/pinmux/Kconfig:55]: OK let's keep this patch in this PR then.

File: soc/arm/nxp_lpc/lpc11u6x/soc.c (hunk 1 above)
Comment (mbittan) [reply soc/arm/nxp_lpc/lpc11u6x/soc.c]: Fixed

File: soc/arm/nxp_lpc/Kconfig [hunk 15]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2017, NXP
+# Copyright (c) 2020, Seagate
```
[the remark this answers is not in the document]
Comment (mbittan) [reply soc/arm/nxp_lpc/Kconfig]: Fixed

File: soc/arm/nxp_lpc/lpc11u6x/Kconfig.soc [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+config SOC_LPC11U66
+	bool "SOC_LPC11U66"
+
+endchoice
+
+if SOC_SERIES_LPC11U6X
+
+config SOC_PART_NUMBER_LPC11U66JBD48
+       bool
```
[the remark this answers is not in the document]
Comment (mbittan) [reply soc/arm/nxp_lpc/lpc11u6x/Kconfig.soc]: Fixed

File: boards/arm/lpcxpresso11u68/Kconfig.defconfig [hunk 17]
```diff
@@ -0,0 +1,11 @@
+# LPCXpresso11U68 board
+
+# Copyright (c) 2020, Seagate
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_LPCXPRESSO11U68
+
+config BOARD
+	default "lpcxpresso11u68" if BOARD_LPCXPRESSO11U68
```
[the remark this answers is not in the document]
Comment (mbittan) [reply boards/arm/lpcxpresso11u68/Kconfig.defconfig]: Fixed

File: soc/arm/nxp_lpc/lpc11u6x/Kconfig.defconfig.series [hunk 18]
```diff
@@ -14,5 +14,11 @@ config NUM_IRQS
 	# must be >= the highest interrupt number used
 	default 40
 
+if CLOCK_CONTROL
+
+config CLOCK_CONTROL_LPC11U6X
+       default y
```
[the remark this answers is not in the document]
Comment (mbittan) [reply soc/arm/nxp_lpc/lpc11u6x/Kconfig.defconfig.series]: Fixed

File: drivers/gpio/Kconfig.lpc11u6x (hunk 6 above)
Comment (mbittan) [reply drivers/gpio/Kconfig.lpc11u6x]: Fixed

File: dts/arm/nxp/nxp_lpc11u6x.dtsi (hunk 7 above)
Comment (mbittan) [reply dts/arm/nxp/nxp_lpc11u6x.dtsi]: Fixed

File: boards/arm/lpcxpresso11u68/lpcxpresso11u68.yaml [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+  - clock_controller
+  - pinmux
+  - gpio
+  - i2c
+  - serial
+  - eeprom
+testing:
+  ignore_tags:
+    - net
+    - bluetooth
```
[the remark this answers is not in the document]
Comment (mbittan) [reply boards/arm/lpcxpresso11u68/lpcxpresso11u68.yaml]: Indeed, they are not and have thus been removed.

File: boards/arm/lpcxpresso11u68/lpcxpresso11u68_defconfig [hunk 20]
```diff
@@ -0,0 +1,21 @@
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=48000000
+CONFIG_GPIO=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_SOC_SERIES_LPC11U6X=y
+CONFIG_MAIN_STACK_SIZE=512
+CONFIG_ISR_STACK_SIZE=768
```
[the remark this answers is not in the document]
Comment (mbittan) [reply boards/arm/lpcxpresso11u68/lpcxpresso11u68_defconfig:9]: Since the board only has 32k of RAM, we decided to lower some of the stack sizes. I added a comment to explain that.

File: boards/arm/lpcxpresso11u68/lpcxpresso11u68_defconfig [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+CONFIG_POLL=y
```
[the remark this answers is not in the document]
Comment (mbittan) [reply boards/arm/lpcxpresso11u68/lpcxpresso11u68_defconfig]: My bad, it shoudn't. It is now fixed.

File: drivers/gpio/Kconfig.lpc11u6x (hunk 13 above)
[the remark this answers is not in the document]
Comment (mbittan) [reply drivers/gpio/Kconfig.lpc11u6x]: We removed the CONFIG_ options and are now using the status to instantiate a GPIO device.

File: drivers/serial/CMakeLists.txt (hunk 9 above)
Comment (mbittan) [reply drivers/serial/CMakeLists.txt]: My bad, we should use it. It is fixed now.

File: drivers/serial/Kconfig.lpc11u6x (hunk 10 above)
Comment (mbittan) [reply drivers/serial/Kconfig.lpc11u6x]: Indeed it can, I removed Kconfig option in favor of device tree macro based on the node's status.

File: drivers/serial/uart_lpc11u6x.c (hunk 11 above)
Comment (mbittan) [reply drivers/serial/uart_lpc11u6x.c]: Indeed, I did not think of that. I implemented the configuration functions for values that are supported by LPC11U6X

File: boards/arm/lpcxpresso11u68/support/lpc_checksum.py (hunk 3 above)
Comment (mbittan) [reply boards/arm/lpcxpresso11u68/support/lpc_checksum.py:3]: @henrikbrixandersen @MaureenHelm @carlescufi We added the README file in boards/arm/lpcxpresso11u68/support needed by the new license process.

File: boards/arm/lpcxpresso11u68/support/lpc_checksum_README [hunk 22]
```diff
@@ -0,0 +1,40 @@
+lpc_checksum
+############
+
+Origin:
+   https://github.com/basilfx/lpc_checksum
+
```
Comment (henrikbrixandersen) [inline boards/arm/lpcxpresso11u68/support/lpc_checksum_README:6]: Instead of including the script within the zephyr repo, could we just add it as an (optional) dependency to be installed using `pip install lpc_checksum`?

File: drivers/gpio/gpio_lpc11u6x.c [hunk 23]
```diff
@@ -0,0 +1,655 @@
...
+GPIO_LPC11U6X_INIT(0);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio1), okay)
+GPIO_LPC11U6X_INIT(1);
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio2), okay)
+GPIO_LPC11U6X_INIT(2);
+#endif
```
Comment (henrikbrixandersen) [inline drivers/gpio/gpio_lpc11u6x.c:655]:
```suggestion
DT_INST_FOREACH_STATUS_OKAY(GPIO_LPC11U6X_INIT)
```

File: drivers/serial/uart_lpc11u6x.c [hunk 24]
```diff
@@ -0,0 +1,996 @@
...
+LPC11U6X_UARTX_INIT(2);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart2), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart3), okay)
+LPC11U6X_UARTX_INIT(3);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart3), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart4), okay)
+LPC11U6X_UARTX_INIT(4);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart4), okay) */
```
Comment (henrikbrixandersen) [inline drivers/serial/uart_lpc11u6x.c:908]:
```suggestion
DT_INST_FOREACH_STATUS_OKAY(LPC11U6X_UARTX_INIT)
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_lpc11u6x.c (hunk 24 above)
Comment (mbittan) [reply drivers/serial/uart_lpc11u6x.c:908]: I'm not sure that would work in my case. On the LPC11U6X MCUs, UART0 has a different interface from the other uart devices, using DT_INST_FOREACH_STATUS_OKAY on LPC11U6X_UARTX_INIT would imply, among other things, initializing all uart devices with the same struct uart_driver_api pointer.
Furthermore, if I understand correctly, there is no guarantee that the instance number Y of a compatible will correspond to the uart Y hardware of the LPC11U6X. In this driver, I need to be sure that the uart device number Y correspond to the actual uart Y of the MCU in order to initialize the shared irq correctly.

File: boards/arm/lpcxpresso11u68/support/lpc_checksum_README (hunk 22 above)
Comment (simonguinot) [reply boards/arm/lpcxpresso11u68/support/lpc_checksum_README:6]: Yes we can :) When I looked in scripts/requirements*, I missed the "optional" part. Is that possible to install the script only when building Zephyr for a LPC11U6x MCU ? It was not feeling right to me to install this script in everybody's $HOME. That's the reason we chose the direct import. But if a dependency is the right way to go, please confirm and I'll do the change.

File: drivers/gpio/gpio_lpc11u6x.c (hunk 23 above)
Comment (simonguinot) [reply drivers/gpio/gpio_lpc11u6x.c:655]: I spent some time this morning trying to do that. The result is I'd rather prefer not :) The way the hardware works, any interruption can be mapped on any GPIO port. So when an interrupt happens, the handler has to retrieve first the GPIO port number and next the "struct device" associated with. To do that, having the GPIO device X associated with the GPIO port X helps a lot. With DT_INST_FOREACH_STATUS_OKAY we can't assume that. If for example the GPIO port 0 node is disabled in DT (let's say by an application overlay), then instance 0 will be GPIO port 1.

File: boards/arm/lpcxpresso11u68/support/lpc_checksum_README (hunk 22 above)
Comment (simonguinot) [reply boards/arm/lpcxpresso11u68/support/lpc_checksum_README:6]: What do you think @henrikbrixandersen ?

Comment (henrikbrixandersen) [reply boards/arm/lpcxpresso11u68/support/lpc_checksum_README:6]: I think we should add the dependency to either `scripts/requirements-base.txt` or `scripts/requirements-extras.txt`.

Comment (mbittan) [reply boards/arm/lpcxpresso11u68/support/lpc_checksum_README]: The dependency has been added to scripts/requirements-extra.txt

File: drivers/serial/uart_lpc11u6x.c (hunk 24 above)
Comment (mbittan) [reply drivers/serial/uart_lpc11u6x.c:908]: @henrikbrixandersen Is there a deterministic way to guarantee that instance number X will correspond to the real uartX of the MCU or should we just keep it like this ?

File: boards/arm/lpcxpresso11u68/support/lpc_checksum.py (hunk 3 above)
Comment (mbittan) [reply boards/arm/lpcxpresso11u68/support/lpc_checksum.py]: lpc_checksum.py has been added as a dependency in the script/requirements-extra.txt file.

File: drivers/serial/uart_lpc11u6x.c (hunk 24 above)
Comment (henrikbrixandersen) [reply drivers/serial/uart_lpc11u6x.c:908]: No, there are no guarantees in that regard. Perhaps @galak has input on how to go about this?

File: drivers/eeprom/eeprom_lpc11u6x.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (chrta) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: I just had a quick look at the IAP of the lpc15xx, see https://www.nxp.com/docs/en/user-guide/UM10736.pdf pg. 572. It seems that the command codes, return values etc are the same. Only the iap entry location seems to be different:
```
#define IAP_LOCATION 0x0300 0205
```

[the remark this answers is not in the document]
Comment (simonguinot) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: > I just had a quick look at the IAP of the lpc15xx, see https://www.nxp.com/docs/en/user-guide/UM10736.pdf pg. 572. It seems that the command codes, return values etc are the same. Only the iap entry location seems to be different:
> 
> ```
> #define IAP_LOCATION 0x0300 0205
> ```

Hi @chrta, Thanks for looking.

So should we move it under soc/arm/nxp_lpc/common right now ? Or shall we rather wait for someone adding support for another LPC SoC needing IAP to do that ?

File: drivers/clock_control/clock_control_lpc11u6x.c [hunk 25]
```diff
@@ -0,0 +1,400 @@
...
+		syscon->pd_run_cfg = (syscon->pd_run_cfg & ~bit)
+			| LPC11U6X_PDRUNCFG_MASK;
+	} else {
+		syscon->pd_run_cfg = syscon->pd_run_cfg | bit
+			| LPC11U6X_PDRUNCFG_MASK;
+	}
+}
+
+static void syscon_set_pll_src(struct lpc11u6x_syscon_regs *syscon,
+			       u32_t src)
```
Comment (MaureenHelm) [inline drivers/clock_control/clock_control_lpc11u6x.c:36]: The zephyr-specific integer types were recently deprecated in #25679. Please rebase and update to use C99 integer types

File: drivers/eeprom/eeprom_lpc11u6x.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: @agansari can you take a look at this?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by agansari — COMMENTED** [review]
need more time to review this pull

File: drivers/eeprom/eeprom_lpc11u6x.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (bbilas) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: > > I just had a quick look at the IAP of the lpc15xx, see https://www.nxp.com/docs/en/user-guide/UM10736.pdf pg. 572. It seems that the command codes, return values etc are the same. Only the iap entry location seems to be different:
> > ```
> > #define IAP_LOCATION 0x0300 0205
> > ```
> 
> Hi @chrta, Thanks for looking.
> 
> So should we move it under soc/arm/nxp_lpc/common right now ? Or shall we rather wait for someone adding support for another LPC SoC needing IAP to do that ?

I have some devboards based on LPC1549 so if you need someone to make some tests and etc I am at your disposal.

File: soc/arm/nxp_lpc/lpc11u6x/iap.h [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+#define IAP_STATUS_INVALID_SECTOR	7
+#define IAP_STATUS_SECTOR_NOT_BLANK	8
+#define IAP_STATUS_SECTOR_NOT_PREPARED	9
+#define IAP_STATUS_COMPARE_ERROR	10
+#define IAP_STATUS_BUSY			11
+
+/**
+ * @brief Entry function for IAP commands.
+ */
+static inline int iap_cmd(unsigned int cmd[5])
```
Comment (agansari) [inline soc/arm/nxp_lpc/lpc11u6x/iap.h:54]: Would it be possible to use already provided functions?
Have a look at modules/hal/nxp/mcux/drivers/lpc/fsl_flashiap.h

File: drivers/eeprom/eeprom_lpc11u6x.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (agansari) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: @mbittan @MaureenHelm the IAP software has the same command structure across older LPC devices; it's used for flash memories most of the time, but in the case if LPC11U6x it's used for eeprom. I see no flash programming in this pull request.
EEPROM driver should be separate from flash, even if they use the same underlying firmware for Zephyr's driver sake, sdk drivers are separate as well.
As bare metal drivers that don't rely on hal_nxp module they are correct and I would approve, but I have not reviewed in detail the other drivers and configurations.
In case flash will be used, a driver that also fixes #6040 and relies on hal/nxp/mcux/drivers/lpc/fsl_flashiap.h would be preferred.

File: drivers/gpio/gpio_lpc11u6x.c [hunk 27]
```diff
@@ -0,0 +1,655 @@
...
+	}
+
+	/*
+	 * Convert the requested GPIO port and pin numbers into a value
+	 * compatible with the INTPIN register (included in the PINTSEL
+	 * register).
+	 */
+	intpin = pin;
+	if (config->port_num == 1) {
+		intpin += 24;
```
Comment (agansari) [inline drivers/gpio/gpio_lpc11u6x.c:354]: This looks similar to https://github.com/zephyrproject-rtos/zephyr/blob/4b194eb4fc40723bba46acf0fae36d5d1d1e59a8/drivers/gpio/gpio_mcux_lpc.c#L199 

Could this be reused?

**Review by agansari — CHANGES_REQUESTED** [review]
I have some questions and remarks.
Looks fine as standalone baremetal drivers, but I think some existing code could be reused from other LPC devices.

File: drivers/gpio/gpio_lpc11u6x.c (hunk 27 above)
Comment (simonguinot) [reply drivers/gpio/gpio_lpc11u6x.c:354]: Hi Andrei,

Many thanks for spending some time on this PR.

In the gpio_mcux_lpc driver each free interrupt can be attached _one_ time to a pin. It is never detached. In all the dtsi files using the "nxp,lpc-gpio" binding, two GPIOs ports are defined and 8 interrupts lines are split equally between this two port. Basically each GPIO port gets four interrupts attached statically at built time. Once the 4 interrupts lines are consumed/used, it is done.

On the LPC11U6x SoC there are 8 interrupts lines available as well, but 3 GPIOs ports. This doesn't divide equally. In addition each of the 8 interrupts can be used by each of the GPIO ports. So we choose a dynamic approach. At run time any of the port can attach any of the free interrupts lines (as longer there is one available). An when an interrupt is disabled then it is detached from the pin. So the interrupt line becomes available again for any of the GPIO ports. It is done by the pintsel_detach function. In the gpio_mcux_lpc driver and in the NXP MCUx HAL, there is no support to detach interrupts from pin. To support the attach/detach feature the pintsel_attach function have to run some checks on the pintsel registers. attach_pin_to_isr() does not. That's why I think that attach_pin_to_isr should not be reused here.

Comment (agansari) [reply drivers/gpio/gpio_lpc11u6x.c:354]: Thank you Simon for your answer. You are right, pins are not changed at runtime, once attached they are not detached, this comes from allocating resources statically in device tree. If you need to detach the pin at runtime it wouldn't be a good ideea to switch to the other driver.
The allocation of 4 interrupt lines/port is arbitrary and should not be taken as hardwired, most of the code in gpio_mcux_lpc is designed so interrupts can be differently distributed depending on application. At the time I implemented that I had LPC55s69 an LPC54114 in mind, the allocation is fair and with no particular application in mind.

File: soc/arm/nxp_lpc/lpc11u6x/iap.h (hunk 26 above)
Comment (agansari) [reply soc/arm/nxp_lpc/lpc11u6x/iap.h:54]: I see your point

File: drivers/serial/uart_lpc11u6x.c (hunk 24 above)
Comment (mbittan) [reply drivers/serial/uart_lpc11u6x.c:909]: @galak Do you have any insight on this ?

File: drivers/eeprom/eeprom_lpc11u6x.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (mbittan) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: @agansari We don't plan on adding support for the flash controller right now, do you think we should keep the EEPROM driver as it is now ?

[the remark this answers is not in the document]
Comment (agansari) [reply drivers/eeprom/eeprom_lpc11u6x.c:7]: @mbittan yes, there is no other LPC device using eeprom with Zephyr support for now. Also:

> EEPROM driver should be separate from flash, even if they use the same underlying firmware for Zephyr's driver sake, sdk drivers are separate as well.

**Review by MaureenHelm — APPROVED** [review]
Will you please investigate the buildkite failure?

**Comment by simonguinot** [comment]
> Will you please investigate the buildkite failure?

Hi Maureen,

Thanks for the approval !

I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by MaureenHelm** [comment]
> I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.

@galak can you help with this?

**Comment by henrikbrixandersen** [comment]
> I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.

@simonguinot You could add it to `install-packages` in `.github/workflows/sanitycheck_tests.yml`.

**Review by agansari — APPROVED** [review]

**Comment by mbittan** [comment]
> > I am not sure how we can fix this failure. Since we have moved the lpc_checksum script into scripts/requirements-extras.txt, then it must be installed in the containers used by the CI.
> 
> @galak can you help with this?

I added a pip3 install command in the .buildkite/run.sh script, as suggested by @galak on slack. It fixes the build failure, but it is only a temporary fix.

**Review by carlescufi — APPROVED** [review]
